### DISCUSSION TITLE
Why is using 'ref' with 'useMemo' in the App Router different from the Page Router when in Strict Mode?

### DISCUSSION THREAD

**Discussion by hawkup** [vercel/next.js#56078] [App Router] [answered]
I'm not sure if I did something wrong or if it's the behavior of both router types, but the log outputs are not the same.

I created a 'useData' hook that sets the 'dataRef' value inside 'useMemo' and added logging before and after. When I use this hook on the App Router and Page Router, the output logs are not the same.

### useData Hook
```tsx
// hooks.ts
import { useRef, useMemo } from "react";

export function useData() {
  const dataRef = useRef<Date | null>(null);

  useMemo(() => {
    if (typeof window !== undefined) {
      const date = new Date();
      console.log(
        "useMemo before set",
        `, dataRef: ${JSON.stringify(dataRef)}`,
        `, date: ${date.toISOString()}`,
      );
      dataRef.current = date;
      console.log(
        "useMemo after set",
        `, dataRef: ${JSON.stringify(dataRef)}`,
        `, date: ${date.toISOString()}`,
      );
    }
  }, []);
}
```

#### App Router
```tsx
// App Router (app/app/page.tsx)
"use client";

import { useData } from "../../hooks";

export default function Home() {
  useData();

  return <div>App Router</div>;
}
```

#### Page Router
```tsx
// Page Router (pages/page.tsx)
import { useData } from "../hooks";

export default function Page() {
  useData();

  return <div>Page Router</div>;
}
```

When looking at the App Router's log, you will notice that during the second rendering, 'dataRef.current' already has the value.
#### App Router's log
![CleanShot 2566-09-27 at 16 39 31](https://github.com/vercel/next.js/assets/2748846/b89839f5-d6a2-4479-b0e9-d8b77258cb05)

#### Page Router's log
Comparing it with the Page Router's log, 'dataRef.current' is null.
![CleanShot 2566-09-27 at 16 46 45](https://github.com/vercel/next.js/assets/2748846/8219c640-bebf-4cf6-9ed8-fe2ee21c40fd)

Here is the link for testing
App router: https://wrvfjp-3000.csb.app/app
Page router: https://wrvfjp-3000.csb.app/page
Code: https://codesandbox.io/p/sandbox/nostalgic-dubinsky-wrvfjp?file=%2Fhooks.ts%3A9%2C19

next version 13.5.3
react version 18.2.0
react-dom version 18.2.0

**Comment by pawelblaszczyk5** [comment]
My guess - it's because the react version between them isn't really the same and there must have been sum internal changes. App router uses canary/experimental React version, while page router uses the one from your `package.json`.

**Comment by icyJoseph** [comment, marked as answer]
Hi,

The partial explanation here is that Next.js during dev mode is using react@canary.

Take this Codesandbox for instance:

- https://codesandbox.io/s/admiring-beaver-fxmtcf?file=/src/App.js

You can see it does as you describe, but it doesn't use Next.js, rather it is just react@canary.

I think you should open an issue in the React repo regarding this, if nobody has reported this already that is.. Probably could even include the Codesanbox above.
